lib/internal/modules/typescript.js

```diff
@@ -0,0 +1,106 @@
+);
+
+/**
+ * Load the TypeScript parser.
+ * and returns an object with a `code` property.
+ * @returns {Function} The TypeScript parser function.
+ */
+const loadTypeScriptParser = getLazy(() => {
+  const amaro = require('internal/deps/amaro/dist/index');
+  return amaro.transformSync;
+});
+
+/**
+ *
+ * @param {string} source the source code
+ * @param {object} options the options to pass to the parser
+ * @returns {TransformOutput} an object with a `code` property.
+ */
+function parseTypeScript(source, options) {
+  const parse = loadTypeScriptParser();
+  return parse(source, options);
+}
+
+function stripTypeScriptTypes(code, options = kEmptyObject) {
+  emitExperimentalWarning('stripTypeScriptTypes');
+  validateObject(options, 'options');
+  const { mode = 'strip-only', sourceMap = false, filename = '' } = options;
+  validateOneOf(mode, 'options.mode', ['strip-only', 'transform']);
+  if (mode === 'strip-only') {
+    validateOneOf(sourceMap, 'options.sourceMap', [false, undefined]);
+    validateOneOf(filename, 'options.filename', ['', undefined]);
+  }
+  validateBoolean(sourceMap, 'options.sourceMap');
+  validateString(filename, 'options.filename');
+
+  const transformOptions = {
+    __proto__: null,
+    mode,
+    sourceMap,
+    filename,
+  };
+
+  const { code: transformed, map } = parseTypeScript(code, transformOptions);
+  if (map) {
+    // TODO(@marco-ippolito) When Buffer.transcode supports utf8 to
+    // base64 transformation, we should change this line.
+    const base64SourceMap = Buffer.from(map).toString('base64');
+    return `${transformed}\n\n//# sourceMappingURL=data:application/json;base64,${base64SourceMap}`;
+  }
+  return transformed;
+}
+
+/**
+ * @typedef {object} TransformOutput
+ * @property {string} code The compiled code.
+ * @property {string} [map] The source maps (optional).
+ *
+ * Performs type-stripping to TypeScript source code.
+ * @param {string} source TypeScript code to parse.
+ * @param {string} filename The filename of the source code.
+ * @returns {TransformOutput} The stripped TypeScript code.
+ */
+function stripTypeScriptModuleTypes(source, filename) {
+  assert(typeof source === 'string');
+  const options = {
+    __proto__: null,
+    mode: getTypeScriptParsingMode(),
+    sourceMap: getOptionValue('--enable-source-maps'),
+    filename,
+  };
+  const { code, map } = parseTypeScript(source, options);
+  if (map) {
+    // TODO(@marco-ippolito) When Buffer.transcode supports utf8 to
+    // base64 transformation, we should change this line.
+    const base64SourceMap = Buffer.from(map).toString('base64');
+    return `${code}\n\n//# sourceMappingURL=data:application/json;base64,${base64SourceMap}`;
+  }
+  // Source map is not necessary in strip-only mode. However, to map the source
+  // file in debuggers to the original TypeScript source, add a sourceURL magic
+  // comment to hint that it is a generated source.
+  return `${code}\n\n//# sourceURL=${filename}`;
```

Comment:
Can't we reduce duplication by reusing `stripTypeScriptTypes`?

```suggestion
  assert(typeof source === 'string');
  const mode = getTypeScriptParsingMode();
  const options = {
    __proto__: null,
    mode,
    sourceMap: getOptionValue('--enable-source-maps'),
    filename,
  };
  const result = stripTypeScriptTypes(source, options);
  // Source map is not necessary in strip-only mode. However, to map the source
  // file in debuggers to the original TypeScript source, add a sourceURL magic
  // comment to hint that it is a generated source.
  if (mode === 'strip-only') {
    return `${result}\n\n//# sourceURL=${filename}`;
  }
```

Reply:
It would perform unnecessary validations, there a few things that can be factorized like adding the sourcemap but other than that they are different